### PR TITLE
Fix: mypy error with `Settings` and `SettingsConfigDict`

### LINKED ISSUE #6985: `pydantic.mypy` plugin error using `pydantic_settings` `SettingsConfigDict`
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

`pydantic` mypy crashes when providing a `model_config` `SettingsConfigDict` within the `BaseSettings` class. `mypy` continues to crash even after commenting out the `model_config` until the virtual environment is rebuilt.

Traceback:
```bash
$ mypy test.py --show-traceback
test.py:3: error: INTERNAL ERROR -- Please try using mypy master on GitHub:
https://mypy.readthedocs.io/en/stable/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 1.4.1
Traceback (most recent call last):
  File "mypy/semanal.py", line 6451, in accept
  File "mypy/nodes.py", line 1139, in accept
  File "mypy/semanal.py", line 1582, in visit_class_def
  File "mypy/semanal.py", line 1667, in analyze_class
  File "mypy/semanal.py", line 1694, in analyze_class_body_common
  File "mypy/semanal.py", line 1777, in apply_class_plugin_hooks
  File "[redacted]/.venv/lib64/python3.8/site-packages/pydantic/mypy.py", line 182, in _pydantic_model_class_maker_callback
    return transformer.transform()
  File "[redacted]/.venv/lib64/python3.8/site-packages/pydantic/mypy.py", line 443, in transform
    self.add_initializer(fields, config, is_settings)
  File "[redacted]/.venv/lib64/python3.8/site-packages/pydantic/mypy.py", line 804, in add_initializer
    base_settings_init_arguments = base_settings_info.names['__init__'].node.arguments
AttributeError: attribute 'names' of 'TypeInfo' undefined
test.py:3: : note: use --pdb to drop into pdb
```

When running the command with `pdb` to print the `base_settings_info` we get the following:
```
-> base_settings_init_arguments = base_settings_info.names['__init__'].node.arguments
(Pdb) base_settings_info
*** AttributeError: attribute '_fullname' of 'TypeInfo' undefined
```

Requirements to reproduce:
```
annotated-types==0.5.0
mypy==1.4.1
mypy-extensions==1.0.0
pydantic==2.1.1
pydantic-core==2.4.0
pydantic-settings==2.0.2
python-dotenv==1.0.0
tomli==2.0.1
typing-extensions==4.7.1
```

mypy config:
```
[tool.mypy]
files = ["api", "tests"]

plugins = [
    "pydantic.mypy"
]
strict = true
```

### Example Code

```Python
from pydantic_settings import BaseSettings, SettingsConfigDict

class Settings(BaseSettings):
    API_HOST: str = "0.0.0.0"

    model_config = SettingsConfigDict(
        env_file=".env", env_file_encoding="utf-8"
    )
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.1.1
        pydantic-core version: 2.4.0
          pydantic-core build: profile=release pgo=true mimalloc=true
                 install path: [REDACTED]/.venv/lib/python3.8/site-packages/pydantic
               python version: 3.8.6 (default, Oct 27 2020, 09:13:12)  [GCC 9.3.1 20200408 (Red Hat 9.3.1-2)]
                     platform: Linux-3.10.0-1160.15.2.el7.x86_64-x86_64-with-glibc2.2.5
     optional deps. installed: ['typing-extensions']
```

Selected Assignee: @Kludex

### PR DESCRIPTION
**Pull request by JeanArhancet**
## Change Summary

Fix mypy error with `Settings` and `SettingsConfigDict`

## Related issue number

Fix #6985

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @davidhewitt

### CODE REVIEW

File: pydantic/mypy.py [hunk 1]
```diff
@@ -800,17 +800,18 @@ def add_initializer(self, fields: list[PydanticModelField], config: ModelConfigD
...
-                settings_init_arguments = []
-                a: Argument
-                for a in base_settings_init_arguments:
-                    if a.variable.name.startswith('__') or not a.variable.name.startswith('_'):
-                        continue
-                    analyzed_variable_type = self._api.anal_type(a.variable.type)
-                    variable = Var(a.variable.name, analyzed_variable_type)
-                    settings_init_arguments.append(Argument(variable, analyzed_variable_type, None, ARG_OPT))
-                args.extend(settings_init_arguments)
+                base_settings_info = self._api.lookup_fully_qualified(BASESETTINGS_FULLNAME).node
```
Comment (hramezani) [inline pydantic/mypy.py:803]:
```suggestion
                base_settings_node = self._api.lookup_fully_qualified(BASESETTINGS_FULLNAME).node
```
```suggestion
                base_settings_info = self._api.lookup_fully_qualified(BASESETTINGS_FULLNAME).node
```

File: pydantic/mypy.py [hunk 2]
```diff
@@ -800,17 +800,18 @@ def add_initializer(self, fields: list[PydanticModelField], config: ModelConfigD
...
+                    base_settings_init_node = base_settings_info.names['__init__'].node
+                    if base_settings_init_node is not None and base_settings_init_node.type is not None:
+                        func_type = base_settings_init_node.type
+                        for arg_idx, arg_name in enumerate(func_type.arg_names):
+                            if arg_name.startswith('__') or not arg_name.startswith('_'):
+                                continue
+                            analyzed_variable_type = self._api.anal_type(func_type.arg_types[arg_idx])
+                            variable = Var(arg_name, analyzed_variable_type)
+                            settings_init_argument = Argument(variable, analyzed_variable_type, None, ARG_OPT)
+                            args.append(settings_init_argument)
```
Comment (hramezani) [inline pydantic/mypy.py:814]:
```suggestion
                            args.append(Argument(variable, analyzed_variable_type, None, ARG_OPT))
```

**Comment by hramezani** [comment]
Thanks @JeanArhancet for the fix 🙏 

This LGTM 👍 

test fast-api is failing that will be fixed after https://github.com/pydantic/pydantic/pull/7014

@dmontagu it would be great if you also take a look

File: tests/mypy/outputs/1.1.1/mypy-plugin_ini/pydantic_settings.py [hunk 3]
```diff
@@ -1,15 +0,0 @@
-from pydantic_settings import BaseSettings
```
Comment (hramezani) [inline tests/mypy/outputs/1.1.1/mypy-plugin_ini/pydantic_settings.py:1]: Do we need to delete this file?

Comment (JeanArhancet) [reply tests/mypy/outputs/1.1.1/mypy-plugin_ini/pydantic_settings.py:1]: I ran the command `test-mypy-update-all` and it destroyed the file

**Comment by martinky24** [comment]
I filed what is believed to be a version of this issue at https://github.com/python/mypy/issues/15804, and after installing this branch it seems to fix the error. :+1:

**Comment by hramezani** [comment]
@JeanArhancet please rebase on top of main.

File: tests/mypy/outputs/1.1.1/mypy-plugin_ini/pydantic_settings.py (hunk 3 above)
Comment (davidhewitt) [reply tests/mypy/outputs/1.1.1/mypy-plugin_ini/pydantic_settings.py:1]: I checked with @dmontagu and this is expected behaviour 👍

**Review by davidhewitt — APPROVED** [review]
Had a quick chat with @dmontagu about this offline, LGTM. Many thanks 👍

**Pull request merged**